itou/www/accounts/views.py

```diff
@@ -0,0 +1,73 @@
+from allauth.account import app_settings
+from allauth.account.models import EmailAddress, EmailConfirmationHMAC
+from allauth.account.utils import send_email_confirmation
+from allauth.account.views import ConfirmEmailView as BaseConfirmEmailView
+from django.conf import settings
+from django.core import signing
+from django.http import Http404
+from django.shortcuts import redirect
+from django.urls import reverse
+from django.utils import timezone
+from django.views.generic import View
+
+from itou.emails.models import Email
+from itou.utils.auth import LoginNotRequiredMixin
+
+
+class ExpiredConfirmationMixin:
+    def get_object_with_expired_key(self):
+        key = self.kwargs["key"]
+        try:
+            pk = signing.loads(key, salt=app_settings.SALT)
+            return EmailConfirmationHMAC(EmailAddress.objects.get(pk=pk, verified=False)).email_address
+        except (
+            signing.SignatureExpired,
+            signing.BadSignature,
+            EmailAddress.DoesNotExist,
+        ):
+            raise Http404()
+
+
+class ConfirmEmailView(BaseConfirmEmailView, ExpiredConfirmationMixin):
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        if self.object is None:
+            try:
+                self.get_object_with_expired_key()  # Raises error for invalid keys.
+                key = self.kwargs["key"]
+            except Http404:
+                key = None
+        context.update({"provided_key": key})
+        return context
+
+
+class ResendConfirmationView(LoginNotRequiredMixin, ExpiredConfirmationMixin, View):
+    def fail(self):
+        return redirect(reverse("signup:choose_user_kind"))
+
+    def get(self, request, *args, **kwargs):
+        # Validate key.
+        try:
+            email_address = self.get_object_with_expired_key()
+        except Http404:
+            return self.fail()
+
+        # Do not send a confirmation to a verified address.
+        if email_address.verified:
+            return self.fail()
+
+        # Rate limiting.
+        this_morning = timezone.now().replace(hour=0, minute=0, second=0, microsecond=0)
+        confirmation_subject_query = "confirmez votre adresse e-mail"
+        if (
+            Email.objects.filter(
+                to__contains=[email_address.email],
+                created_at__gte=this_morning,
+                subject__icontains=confirmation_subject_query,
+            ).count()
+            >= settings.ACCOUNT_MAX_DAILY_EMAIL_CONFIRMATION_REQUESTS
+        ):
+            return redirect(reverse("account_email_rate_limit_exceeded"))
```

Comment:
Humpf. On a Redis pour implémenter des rate limiting. Demander à la DB de filtrer sur une table avec des millions d’entrées me semble coûteux et complexe, se baser sur le sujet de l’email me semble très instable.

Reply:
En reflexion t'as raison, je vais suggerer une nouvelle solution en utilisant le cache. Je vois que le cache est utilisé par des autres solutions rate limiting et au bout d'un jour, ces données ne sont plus intéressantes. Je suis sous l'impression qu'on a bien besoin de logique sur mesure pour ce besoin.

J'avais ces deux concernes, je pensais qu'avec l'indexe sur le champ `to` ce ne serait pas si couteux mais je n'ai pas testé. Autrement le test va se planter si le contenu du sujet change, je pense que cette partie est stable en practique (même si c'est un peu hacky).